packages/rest/src/client/AbstractRestClient.ts

```diff
@@ -365,39 +372,98 @@ export abstract class AbstractRestClient {
+     * @param {*} [nodeClientError] - If the source is a node http client error (meaning the request
+     *                                did not make it to the remote system) this parameter should be
+     *                                populated.
+     * @returns {RestClientError} - The error that can be thrown or rejected.
+     */
+    private populateError(error: IRestClientError, nodeClientError?: any): RestClientError {
+
+        // Final error object parameters
+        let finalError: IRestClientError = error;
+
+        // @deprecated - extract the status code - now moved to HTTP status.
+        const errorCode = this.response == null ? undefined : this.response.statusCode;
+
+        // start off by coercing the request details to string in case an error is encountered trying
+        // to stringify / inspect them
+        let headerDetails: string = this.mReqHeaders + "";
+        let payloadDetails: string = this.mWriteData + "";
+        try {
+            headerDetails = JSON.stringify(this.mReqHeaders);
+            payloadDetails = inspect(this.mWriteData, { depth: null });
+        } catch (stringifyError) {
+            this.log.error("Error encountered trying to parse details for REST request error:\n %s", inspect(stringifyError, { depth: null }));
+        }
+
+        // Populate the "relevant" fields - caller will have the session, so
+        // no need to duplicate "everything" here, just host/port for easy diagnosis
+        finalError.errorCode = errorCode;
+        finalError.port = this.mSession.ISession.port;
+        finalError.host = this.mSession.ISession.hostname;
+        finalError.basePath = this.mSession.ISession.basePath;
+        finalError.httpStatus = errorCode;
+        finalError.errno = (nodeClientError != null) ? nodeClientError.errno : undefined;
+        finalError.syscall = (nodeClientError != null) ? nodeClientError.syscall : undefined;
+        finalError.payload = this.mWriteData;
+        finalError.headers = this.mReqHeaders;
+        finalError.resource = this.mResource;
+        finalError.request = this.mRequest;
+
+        // Construct a formatted details message
+        const detailMessage: string =
+        ((finalError.source === "client") ?
+        `HTTP(S) client encountered an error. Request could not be initiated to host.\n` +
+        `Review connection details (host, port) and ensure correctness.`
+        :
+        `HTTP(S) error status "${finalError.httpStatus}" received.\n` +
+        `Review request details (resource, base path, credentials, payload) and ensure correctness.`) +
+        "\n" +
+        "\nHost:      " + finalError.host +
+        "\nPort:      " + finalError.port +
+        "\nBase Path: " + finalError.basePath +
+        "\nResource:  " + finalError.resource +
+        "\nRequest:   " + finalError.request +
+        "\nHeaders:   " + headerDetails +
+        "\nPayload:   " + payloadDetails;
```

Comment:
This increased amount of information should hopefully be helpful to the user :+1: